### PR TITLE
RATIS-1861. NullPointerException in readAsync when Ratis leader is changing

### PR DESCRIPTION
**Pull request by BUAAserein**
see [jira1861](https://issues.apache.org/jira/browse/RATIS-1861).

### CODE REVIEW

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Review by SzyWilliam — APPROVED** [review]
+1 the changes look good to me. Thanks @BUAAserein for fixing this!

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java [hunk 1]
```diff
@@ -597,10 +597,11 @@ synchronized void changeToLeader() {
     Preconditions.assertTrue(getInfo().isCandidate());
     role.shutdownLeaderElection();
     setRole(RaftPeerRole.LEADER, "changeToLeader");
+    role.updateLeaderState(this);
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:600]: Let's store the LeaderStateImpl and then use it later.
```java
@@ -603,10 +603,11 @@ class RaftServerImpl implements RaftServer.Division,
     Preconditions.assertTrue(getInfo().isCandidate());
     role.shutdownLeaderElection();
     setRole(RaftPeerRole.LEADER, "changeToLeader");
+    final LeaderStateImpl leader = role.updateLeaderState(this);
     state.becomeLeader();
 
     // start sending AppendEntries RPC to followers
-    final LogEntryProto e = role.startLeaderState(this);
+    final LogEntryProto e = leader.start();
     getState().setRaftConf(e);
   }
```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RoleInfo.java [hunk 2]
```diff
@@ -79,8 +79,12 @@ LeaderStateImpl getLeaderStateNonNull() {
...
   }
 
-  LogEntryProto startLeaderState(RaftServerImpl server) {
-    return updateAndGet(leaderState, new LeaderStateImpl(server)).start();
+  void updateLeaderState(RaftServerImpl server) {
+    updateAndGet(leaderState, new LeaderStateImpl(server));
+  }
+
+  LogEntryProto startLeaderState() {
+    return leaderState.get().start();
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/RoleInfo.java:87]: Then, we don't need the new `startLeaderState()` method.
```java
+++ b/ratis-server/src/main/java/org/apache/ratis/server/impl/RoleInfo.java
@@ -79,8 +79,8 @@ class RoleInfo {
     return Objects.requireNonNull(leaderState.get(), "leaderState is null");
   }
 
-  LogEntryProto startLeaderState(RaftServerImpl server) {
-    return updateAndGet(leaderState, new LeaderStateImpl(server)).start();
+  LeaderStateImpl updateLeaderState(RaftServerImpl server) {
+    return updateAndGet(leaderState, new LeaderStateImpl(server));
   }
 
```j

**Review by szetszwo — COMMENTED** [review]
@BUAAserein , thanks a lot for working on this!  Please see the minor comments inlined.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java (hunk 1 above)
Comment (BUAAserein) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java]: OK

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RoleInfo.java (hunk 2 above)
Comment (BUAAserein) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/RoleInfo.java]: Done. Thanks a lot for the comments!

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**